toolchain/sem_ir/inst_profile.cpp

```diff
@@ -11,113 +11,108 @@
-
-// If ArgT is IntId, then profile the `APInt`, which profiles the bit-width and
-// value.
-// TODO: Should IntIds themselves be deduplicated?
-template <>
-constexpr ProfileArgFunction* SelectProfileArgTypeFunction<IntId> =
-    IntProfileArgFunction;
-
-// If ArgT is RealId, then profile the value.
-// TODO: Should RealIds themselves be deduplicated?
-template <>
-constexpr ProfileArgFunction* SelectProfileArgTypeFunction<RealId> =
-    RealProfileArgFunction;
+static constexpr auto SelectProfileArgFunction() -> ProfileArgFunction* {
+  if constexpr (N >= InstLikeTypeInfo<InstT>::NumArgs) {
+    // This argument is not used by this instruction; don't profile it.
+    return NullProfileArgFunction;
+  } else {
+    using ArgT = typename InstLikeTypeInfo<InstT>::template ArgType<N>;
+    if constexpr (std::is_same_v<ArgT, InstBlockId>) {
+      return InstBlockProfileArgFunction;
+    } else if constexpr (std::is_same_v<ArgT, TypeBlockId>) {
+      return TypeBlockProfileArgFunction;
+    } else if constexpr (std::is_same_v<ArgT, IntId>) {
+      return IntProfileArgFunction;
+    } else if constexpr (std::is_same_v<ArgT, RealId>) {
+      return RealProfileArgFunction;
+    } else {
+      return DefaultProfileArgFunction;
+    }
+  }
+}
 
 // Profiles the given instruction arguments using the specified functions.
 template <ProfileArgFunction* ProfileArg0, ProfileArgFunction* ProfileArg1>
-static auto ProfileArgs(llvm::FoldingSetNodeID& id, File* context, int32_t arg0,
-                        int32_t arg1) -> void {
-  ProfileArg0(id, context, arg0);
-  ProfileArg1(id, context, arg1);
-}
-
-static constexpr auto KindIndex(InstKind::RawEnumType kind) -> int {
-  return static_cast<InstKind::UnderlyingType>(kind);
+static auto ProfileArgs(llvm::FoldingSetNodeID& id, const File& sem_ir,
+                        int32_t arg0, int32_t arg1) -> void {
+  ProfileArg0(id, sem_ir, arg0);
+  ProfileArg1(id, sem_ir, arg1);
 }
 
-auto ProfileConstant(llvm::FoldingSetNodeID& id, File* sem_ir, Inst inst)
+auto ProfileConstant(llvm::FoldingSetNodeID& id, const File& sem_ir, Inst inst)
     -> void {
   using ProfileArgsFunction =
-      auto(llvm::FoldingSetNodeID&, File*, int32_t, int32_t)->void;
+      auto(llvm::FoldingSetNodeID&, const File&, int32_t, int32_t)->void;
   static constexpr ProfileArgsFunction* ProfileFunctions[] = {
-#define CARBON_SEM_IR_INST_KIND(KindName)            \
-  ProfileArgs<SelectProfileArgFunction<KindName, 0>, \
-              SelectProfileArgFunction<KindName, 1>>,
+#define CARBON_SEM_IR_INST_KIND(KindName)              \
+  ProfileArgs<SelectProfileArgFunction<KindName, 0>(), \
+              SelectProfileArgFunction<KindName, 1>()>,
 #include "toolchain/sem_ir/inst_kind.def"
   };
 
   inst.kind().Profile(id);
-  ProfileFunctions[KindIndex(inst.kind())](id, sem_ir, inst.arg0(),
-                                           inst.arg1());
+  ProfileFunctions[inst.kind().AsInt()](id, sem_ir, inst.arg0(), inst.arg1());
```

Comment:
The refactoring made me think to suggest a different approach:

```
static constexpr std::pair<ProfileArgFunction*, ProfileArgFunction*>
    ProfileArgFunctions[] = {
#define CARBON_SEM_IR_INST_KIND(KindName)   \
  {SelectProfileArgFunction<KindName, 0>(), \
   SelectProfileArgFunction<KindName, 1>()},
#include "toolchain/sem_ir/inst_kind.def"
};

inst.kind().Profile(id);
auto functions = ProfileArgFunctions[inst.kind().AsInt()];
functions.first(id, sem_ir, inst.arg0());
functions.second(id, sem_ir, inst.arg1());
```

I think this eliminates the `ProfileArgs` function.

Totally optional -- not sure what you prefer.

Reply:
FWIW, I started with that, but switched to this approach so that we make only one indirect branch instead of two. Honestly I'm not sure which is better; I can see arguments in both directions, but I *think* the single indirect branch approach will likely be more efficient.